addons/udes_stock/models/stock_picking.py

```diff
@@ -3093,3 +3098,43 @@ def open_back_orders(self):
         return self.open_related_pickings(
             self.u_created_back_orders.ids, "Created Back Orders"
         )
+
+    @api.multi
+    def safe_prevent_date_done_on_not_done_pickings(self):
+        """
+        Set date_done field to empty in case that not all moves are in state done
+        """
+        for picking, moves in self.mapped("move_lines").groupby("picking_id"):
```

Comment:
Q: We already have the pickings in self, why do we need to group by the move lines then by picking again?

We could instead loop through the pickings filtering those with date_done set, then do a `search_count` for moves of that picking ID that are not done, and if there are any, clear the date.

Reply:
self.safe_prevent_date_done_on_not_done_pickings(), self can be multiple pickings, as i am understanding from  action_done method, needed to handle that scenario i think

Reply:
Yes, so `self` with have multiple pickings. So we can do `for picking in self` - then check the moves from there right?

Reply:
Or probably better to do `for picking in self.filtered("date_done")` for example

Reply:
```
for picking in self.filtered("date_done"):
    open_move_count = Move.search_count([("picking_id", "=", picking.id), ("state", "!=", "done")])
    if open_move_count:
        # clear picking done date
```

Something like this? It would avoid needing to group or reference all moves, which on big systems can have an impact on performance.

Reply:
Not agree with this as maybe that some moves are on state cancel, and we dont want to remove date_done on those, as @PeterAlabaster suggested which i agree should check that at least one should be done.
Pushed another change for that if happy.
` open_move_count = Move.search_count([("picking_id", "=", picking.id), ("state", "!=", "done")])`

Agree about mapping and groupping

Reply:
The reason why i did that is because i wanted moves to and instead assigning them again like moves = picking.move_lines, thought better to group all moves by picking id

Reply:
agreed, but I think we could at least remove the groupby, just do `for picking in self: moves = picking.mapped("move_line_ids")`

Reply:
We should avoid `mapped` here, on big pickings it will be ineffective.

Reply:
Didn't consider cancelled moves.

Maybe:

```
Move = self.env["stock.move"].with_context(prefetch_fields=False)
...
for picking in self.filtered("date_done"):
    clear_date_done = False
    none_done_moves = Move.search([("picking_id", "=", picking.id), ("state", "!=", "done")], order="id")
    if none_done_moves:
        none_done_moves.read(["state"])
        for move in none_done_moves:
            if move.state != "cancel":
                clear_date_done = True
                break
    if clear_date_done:
        picking.write({"date_done": False})
```

Reply:
Or we could search for all moves of any picking in self first, then group by those by picking ID and then we still avoid mapped. That would avoid calling `search` multiple times if `self` has more than 1 picking.

Reply:
So we could do:

```
none_done_moves = Move.search([("picking_id", "in", self.filtered("date_done").ids), ("state", "!=", "done")], order="id")

for picking, move in none_done_moves.groupby("picking_id"):
    ...
```

I don't really mind too much with this, just the `mapped` calls we should avoid if possible, as we've seen many times where this has been a bit of a bottleneck.

Reply:
If we filter out the moves of a pick that are done, and if there are cancel moves in the same picking we maybe empty the field date_done where we shouldnt. I think we shouldnt filter out the moves that are done